### PR TITLE
Use correct tree for child unapply context

### LINKED ISSUE #12182: Extraction spoils generic type erasure
The following code produces a ClassCastException at runtime:

```scala
object Test {
  abstract class Base[T] {
    def v: T
  }
  class ImplI extends Base[Int] {
    def v = 42
  }
  object ImplI {
    def unapply(x: ImplI) = Some(x.v)
  }
  class ImplD extends Base[Double] {
    def v = 42.0
  }
  object ImplD {
    def unapply(x: ImplD) = Some(x.v)
  }
  def foo[T](x: Base[T], labels: Seq[T]): Unit = {
    val last: T = labels.last // ClassCastException

    x match {
      // uncomment to fix
      // case ImplI(v) => println(v)
      case ImplD(v) => println(v)
      case ImplI(v) => println(v)
    }
  }
  def main(args: Array[String]): Unit = {
    foo(new ImplI(), Seq(1, 2, 3))
  }
}
```

Mathing `ImplI(_)` before `ImplD(_)` as well as not matching `ImplD(_)` at all fixes the failure.

Probably related to #8563.

### LINKED ISSUE #12194: Custom extractor causes wrong type to be inferred in different match block
## reproduction steps

using Scala 2.13.3 on JDK 14

```scala
object Main {

  trait Trait[A]

  class CaseA extends Trait[Int]

  class CaseB extends Trait[Boolean]

  object Pattern {
    def unapply(fa: CaseB): Option[CaseB] = None
  }

  def foo[A](t: Trait[A]) = {
    def bar(f: Trait[A]): Unit = {}

    t match {
      case Pattern(_) =>
    }

    t match {
      case a: CaseA => bar(new CaseA)
    }
  }

}

```
## problem

Both match-blocks compile fine on their own, with both blocks compilation fails with:
```
Foo.scala:21:28: type mismatch;
  found   : Main.CaseA
  required: Main.Trait[A]
```

In the second one, the type parameter `A` shoud be inferred to be `Int`, which allows a new instance of `CaseA` to be passed to bar. This works fine if the first match block is removed.

However adding the first match block breaks that. A completely different block should of course not have any effect on local inferrence at all.

The issue does not happen if instead of the extractor a type-test for `CaseB` is used or `unapply` expects `Trait[_]`, `Any` or `CaseA`. Also it is important that `CaseA` and `CaseB` extend `Trait[A]` with different type arguments. It seems that the local inferrence of the first influences the second match.

### PR DESCRIPTION
**Pull request by joroKr21**
Prevents `ClassCastException` in some GADT scenarios
and makes valid code compile in others.

### Problem

The parent context tree could be the case def itself.
In that case creating a child context with the same tree
breaks the logic to restore saved type bounds in `typedCase`,
because `enclosingCaseDef` returns the child instead of the parent.
This leads to saving modified bounds in the child context
but later trying and failing to restore them in the parent context.

The leaked bounds later cause type errors in the best case
or propagate incorrect information to erasure in the worst case.

### Solution

Using the tree of the unapply itself for the child context
means that `enclosingCaseDef` will return the parent context.

fixes scala/bug#12194
fixes scala/bug#12182

### CODE REVIEW

**Comment by som-snytt** [comment]
That does seem wiser than using the incorrect tree.

**Comment by joroKr21** [comment]
Parents sit at the parents' table, children sit at the children's table.

File: src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala [hunk 1]
```diff
@@ -311,7 +311,7 @@ trait PatternTypers {
               }
 
               val GenPolyType(freeVars, unappFormal) = freshArgType(unapplyType.skolemizeExistential(context.owner, tree))
-              val unapplyContext = context.makeNewScope(context.tree, context.owner)
+              val unapplyContext = context.makeNewScope(tree, context.owner)
```
Comment (joroKr21) [inline src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala:314]: -8 characters fix ❗

**Comment by joroKr21** [comment]
I just had another laugh at the expense of this bug 😆 
https://github.com/scala/scala/blob/2.13.x/src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala#L266-L270
```scala
      // use "tree" for the context, not context.tree: don't make another CaseDef context,
      // as instantiateTypeVar's bounds would end up there
      val ctorContext = context.makeNewScope(tree, context.owner)
      freeVars foreach ctorContext.scope.enter
      newTyper(ctorContext).infer.inferConstructorInstance(tree1, undetparams, ptSafe)
```

**Review by dwijnand — APPROVED** [review]
❤️ 💪

**Pull request merged**